Xamarin.Essentials/FilePicker/FilePicker.ios.cs

```diff
@@ -0,0 +1,132 @@
+            var documentPicker = new UIDocumentPickerViewController(allowedUtis, UIDocumentPickerMode.Import);
+
+            documentPicker.DidPickDocumentAtUrls += DocumentPicker_DidPickDocumentAtUrls;
+            documentPicker.DidPickDocument += DocumentPicker_DidPickDocument;
+            documentPicker.WasCancelled += DocumentPicker_WasCancelled;
+
+            var parentController = Platform.GetCurrentViewController();
+
+            parentController.PresentViewController(documentPicker, true, null);
+
+            var tcs = new TaskCompletionSource<PickResult>();
+
+            Interlocked.Exchange(ref completionSource, tcs);
+
+            return tcs.Task;
+        }
+
+        static void DocumentPicker_DidPickDocumentAtUrls(object sender, UIDocumentPickedAtUrlsEventArgs args)
+        {
+            // this is called starting from iOS 11.
+            var control = (UIDocumentPickerViewController)sender;
+            foreach (var url in args.Urls)
+            {
+                DocumentPicker_DidPickDocument(control, new UIDocumentPickedEventArgs(url));
+            }
+
+            control.Dispose();
+        }
+
+        static void DocumentPicker_DidPickDocument(object sender, UIDocumentPickedEventArgs args)
+        {
+            try
+            {
+                var securityEnabled = args.Url.StartAccessingSecurityScopedResource();
+                var doc = new UIDocument(args.Url);
+
+                string filename = doc.LocalizedName;
+                string pathname = doc.FileUrl?.Path;
+
+                args.Url.StopAccessingSecurityScopedResource();
+
+                // iCloud drive can return null for LocalizedName.
+                if (filename == null && pathname != null)
+                {
+                    filename = Path.GetFileName(pathname);
+                }
+
+                var result = new PickResult(filename, pathname);
+                var tcs = Interlocked.Exchange(ref completionSource, null);
+                tcs?.SetResult(result);
+            }
+            catch (Exception ex)
+            {
+                // pass exception to task so that it doesn't get lost in the UI main loop
+                var tcs = Interlocked.Exchange(ref completionSource, null);
+                tcs.SetException(ex);
+            }
+        }
+
+        static void DocumentPicker_WasCancelled(object sender, EventArgs args)
+        {
+            var tcs = Interlocked.Exchange(ref completionSource, null);
+            tcs.SetResult(null);
+        }
+    }
+
+    public partial class PickOptions
+    {
+        static PickOptions PlatformGetImagesPickOptions()
+        {
+            return new PickOptions
+            {
+                FileTypes = new string[] { UTType.Image }
+            };
+        }
+    }
+
+    public partial class PickResult
+    {
+        Stream stream;
+
+        string PlatformFileUri { get; set; }
+
+        string PlatformFileName { get; set; }
```

Comment:
I don't think we need to have the two layers of properties. Just expose the `FileName` and `FileUri` properties directly here and set it via the constructor. You can make the setter either `protected` or `internal`

Reply:
And done.